[Clang] [NFC] Expansion Statements (Part 4: for-range refactor)

Sirraide opened the pull request:
This is an NFC patch that refactors the range-based `for` code such that it can be reused for iterating expansion statements. It also adds a `ForExpansionStmt` bool parameter to some functions, but this is still always `false` and currently unused (it will be used by the next patch in the series).

Sirraide committed:
[Clang] [C++26] Expansion Statements (Part 4)

Sirraide committed:
[Clang] [C++26] Expansion Statements (Part 4)

Sirraide commented:
Update: Iterating expansion statements require building a lambda in Sema (see part 5 in this patch series), which necessitated moving `ParseScope` and friends from the Parser into Sema; I’ve included this refactor in this NFC patch.

As an aside, it makes more sense for `ParseScope` etc. to be in Sema anyway since its implementation exclusively calls Sema functions—technically, it does call _some_ parser functions, but _those functions_ then in turn only call Sema functions and don’t actually access any parser state. The only reference to parser state is that the current token location is passed to Sema... but Sema doesn’t do anything w/ it, so I just removed that parameter.

cor3ntin reviewed (commented):
I'm a bit uncomfortable moving parser related objects to sema. @AaronBallman opinions?

Sirraide commented:
> I'm a bit uncomfortable moving parser related objects to sema.

Well I would argue it’s ‘parser-related’ in name only; what it does is it pushes a new scope; the implementation of `ParseScope` only calls Sema functions; it never actually did anything `Parser`-related in the first place

Sirraide commented:
> the implementation of `ParseScope` only calls Sema functions

(yes, it does call e.g. `Parser::EnterScope`, but that function in turn only calls Sema functions and doesn’t modify any state of its `Parser` object; and the same applies to all the other `Parser` functions that are called by members of `ParseScope`)

AaronBallman commented:
> > I'm a bit uncomfortable moving parser related objects to sema.
> 
> Well I would argue it’s ‘parser-related’ in name only; what it does is it pushes a new scope; the implementation of `ParseScope` only calls Sema functions; it never actually did anything `Parser`-related in the first place

To me, the `ParseScope` object is an RAII one that should live in `RAIIObjectsForParser.h`; it exists for `Parser` and `Sema` to collude on scope management from the perspective of the parser. (e.g., when parsing, you know it's time to enter a new scope so you'd use one of these to manage it automagically for you). Based on that, it seems like this interface should remain at the `Parser` level; Sema should not need to enter new parse scopes in general *except* in the case of doing statement rewriting which should be an exception rather than a regular thing, right? (IOW, I think we'd want to go in the direction of generic facilities to support statement rewriting rather than lower-level facilities like the parser has to use?)

This isn't a strongly held opinion btw.

cor3ntin commented on clang/lib/Sema/SemaStmt.cpp:
```diff
@@ -2431,27 +2431,46 @@ void NoteForRangeBeginEndFunction(Sema &SemaRef, Expr *E,
+StmtResult Sema::BuildCXXForRangeRangeVar(Scope *S, Expr *Range, QualType Type,
+                                          bool Constexpr) {
+
+  // Divide by 2, since the variables are in the inner scope (loop body).
+  const auto DepthStr = std::to_string(S->getDepth() / 2);
+  IdentifierInfo *Name =
+      PP.getIdentifierInfo(std::string("__range") + DepthStr);
```
I really think we should not name these things - that being said it's pre-existing

Sirraide replied:
Not naming them might end up generating weird diagnostics, but I’m not sure it’s a problem w/ this particular variable (see https://github.com/llvm/llvm-project/pull/169685#discussion_r3493897265 for more information)

shafik commented on clang/lib/Sema/SemaStmt.cpp:
```diff
@@ -2741,208 +2964,36 @@ StmtResult Sema::BuildCXXForRangeStmt(
-      return StmtError();
-    QualType RangeType = Range->getType();
+    auto BeginRangeRefTy = RangeVar->getType().getNonReferenceType();
+    auto [BeginVar, EndVar, BeginExpr, EndExpr] = BuildCXXForRangeBeginEndVars(
```
I am conflicted here, you made `ForRangeBeginEndInfo` and aggregate but you gave it an `isValid` member but then just decompose it here and that member has no value. 

This is the problem w/ aggregates, they are nice and convenient to use but if you really do want to do checking then we should do proper checking.

These types have a way to ending up being used more widely later on and then the sloppiness comes backs and bites us. 

I would advocate against using it as an aggregate and be stricter here.

Sirraide replied:
In this function, we just manually check whether `BeginVar`/`EndVar` are null, which is equivalent to checking `isValid()`.

shafik replied:
My issue is long term maintenance. It happens a simple limited use struct becomes more widely used and what was a quick trade-off becomes a burden. I would prefer to just use the member function and not decompose.

shafik commented on clang/lib/Sema/SemaStmt.cpp:
```diff
@@ -2431,27 +2431,46 @@ void NoteForRangeBeginEndFunction(Sema &SemaRef, Expr *E,
-  VarDecl *Decl = VarDecl::Create(SemaRef.Context, DC, Loc, Loc, II, Type,
-                                  TInfo, SC_None);
+VarDecl *Sema::BuildForRangeVarDecl(SourceLocation Loc, QualType Type,
+                                    IdentifierInfo *II, bool Constexpr) {
```
Is `Constexpr` just not used, then why pass it?

Sirraide replied:
It will be used in patch 5

shafik commented on clang/lib/Sema/SemaStmt.cpp:
```diff
@@ -2710,6 +2715,224 @@ static StmtResult RebuildForRangeWithDereference(Sema &SemaRef, Scope *S,
+    Scope *S, VarDecl *RangeVar, SourceLocation ColonLoc,
+    SourceLocation CoawaitLoc,
+    ArrayRef<MaterializeTemporaryExpr *> LifetimeExtendTemps,
+    BuildForRangeKind Kind, bool Constexpr, StmtResult *RebuildResult,
```
Does `Constexpr` mean "is constexpr"? The name feels ambiguous.

Sirraide replied:
I’m not sure what else it could mean but I can rename it to `IsConstexpr` if that would be clearer

shafik replied:
Especially since we use this elsewhere. That is why I asked since we use `IsConstexpr` all over and I was puzzled why we just did not stick w/ that.

shafik commented on clang/lib/Sema/SemaStmt.cpp:
```diff
@@ -2431,27 +2431,46 @@ void NoteForRangeBeginEndFunction(Sema &SemaRef, Expr *E,
   SemaRef.Diag(Loc, diag::note_for_range_begin_end)
     << BEF << IsTemplate << Description << E->getType();
 }
+}
```
Why do we have this new `}` are we ending a namespace earlier? It should have a comment indicating the name of the namespace it is closing. This one looks like it should be `// unnamed namespace`

Sirraide replied:
I think that should happen automatically when I run `clang-format`, which I will do once all the patches have been merged into one patch

Sirraide replied:
On that note the brace didn’t have a comment before either (if you look at line 2448 in the diff, you’ll see a brace that was ‘deleted’ there when really it was just moved here)

shafik replied:
We should be adding these comments regardless: https://llvm.org/docs/CodingStandards.html#namespace-indentation

I don't think clang-format does that but maybe I never realized it.

shafik commented on clang/lib/Sema/SemaStmt.cpp:
```diff
@@ -2710,6 +2715,224 @@ static StmtResult RebuildForRangeWithDereference(Sema &SemaRef, Scope *S,
+    return {};
+
+  // Build auto __begin = begin-expr, __end = end-expr.
+  // Divide by 2, since the variables are in the inner scope (loop body).
```
So I see this comment twice, this feels like duplicate code that is waiting for a diverging bugs to happen with over time. Can we consolidate this process into some sort of helper function. Maybe something to generate these names?

Sirraide replied:
This is pre-existing code; we could definitely come up w/ a function for this, but I feel like that is out of scope for this patch series

shafik replied:
Since this is well known code smell and we are touching this code we should fix it. I would be ok if you create a task and just did it as a follow-up.

Sirraide replied:
Opened #208770 to figure out what to do about this.

shafik commented on clang/lib/Sema/SemaStmt.cpp:
```diff
@@ -2710,6 +2715,224 @@ static StmtResult RebuildForRangeWithDereference(Sema &SemaRef, Scope *S,
+    } else {
+      // Can't be a DependentSizedArrayType or an IncompleteArrayType since
+      // UnqAT is not incomplete and Range is not type-dependent.
+      llvm_unreachable("Unexpected array type in for-range");
```
Why can't we just diagnose?

Sirraide replied:
This is pre-existing code; I just moved it around

shafik replied:
Do we know if there is a reason we just don't diagnose?

Sirraide replied:
No idea; presumably it really just can’t happen

shafik commented on clang/lib/Sema/SemaStmt.cpp:
```diff
@@ -2710,6 +2715,224 @@ static StmtResult RebuildForRangeWithDereference(Sema &SemaRef, Scope *S,
+    if (const ConstantArrayType *CAT = dyn_cast<ConstantArrayType>(UnqAT))
+      BoundExpr = IntegerLiteral::Create(
+          Context, CAT->getSize(), Context.getPointerDiffType(), RangeLoc);
+    else if (const VariableArrayType *VAT =
```
Could we just make this ill-formed for now and deal w/ it later? This looks problematic and something we could tackle later on.

CC @AaronBallman

Sirraide replied:
Likewise, this is not new code; I just moved it around

shafik reviewed (commented):
I am a bit surprised no tests were changed, it looks like there are ton more error paths and so I would expect this to be reflected in a test someplace.

Sirraide committed:
[Clang] [C++26] Expansion Statements (Part 5: Iterating Expansion Statements) (#169684)

This implements Sema and template instantiation for iterating expansion
statements, as well as expansion statements whose initialiser is
dependent (and which could either be iterating or destructuring).

Currently, actually expanding these is not supported; this will be done
in a follow-up patch.
